File: internal/stack/stack.go
package stack

import (
	"bufio"
	"errors"
	"fmt"
	"os"
	"strings"
)

var ErrEmptyStack = errors.New("directory stack is empty")

// Stack handles the directory stack for push/pop operations
type Stack struct {
	path string
}

// New creates a new Stack instance
func New(path string) *Stack {
	return &Stack{path: path}
}

// Push adds a directory to the top of the stack
func (s *Stack) Push(dir string) error {
	// Read existing stack
	entries, err := s.load()
	if err != nil {
		return err
	}

	entries = append(entries, dir)
	return s.save(entries)
}

// Pop removes and returns the top directory from the stack
func (s *Stack) Pop() (string, error) {
	entries, err := s.load()
	if err != nil {
		return "", err
	}

	if len(entries) == 0 {
		return "", ErrEmptyStack
	}

	// Get last entry (top of stack)
	dir := entries[len(entries)-1]
	entries = entries[:len(entries)-1]

	if err := s.save(entries); err != nil {
		return "", err
	}

	return dir, nil
}

// Peek returns the top directory without removing it
func (s *Stack) Peek() (string, error) {
	entries, err := s.load()
	if err != nil {
		return "", err
	}

	if len(entries) == 0 {
		return "", ErrEmptyStack
	}

	return entries[len(entries)-1], nil
}

// Size returns the number of entries in the stack
func (s *Stack) Size() (int, error) {
	entries, err := s.load()
	if err != nil {
		return 0, err
	}
	return len(entries), nil
}

// Clear removes all entries from the stack
func (s *Stack) Clear() error {
	return s.save([]string{})
}

func (s *Stack) load() ([]string, error) {
	file, err := os.Open(s.path)
	if os.IsNotExist(err) {
		return []string{}, nil
	}
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var entries []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line != "" {
			entries = append(entries, line)
		}
	}

	return entries, scanner.Err()
}

func (s *Stack) save(entries []string) error {
	file, err := os.Create(s.path)
	if err != nil {
		return err
	}
	defer file.Close()

	for _, entry := range entries {
		if _, err := fmt.Fprintln(file, entry); err != nil {
			return err
		}
	}

	return nil
}

File: src/stack.rs
//! Directory stack for push/pop navigation

use std::fs::{self, File};
use std::io::{BufRead, BufReader, Write};
use std::path::PathBuf;

use thiserror::Error;

/// Errors that can occur during stack operations
#[derive(Error, Debug)]
pub enum StackError {
    #[error("directory stack is empty")]
    Empty,

    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),
}

/// Directory stack for push/pop operations
pub struct Stack {
    path: PathBuf,
}

impl Stack {
    pub fn new(path: PathBuf) -> Self {
        Self { path }
    }

    /// Push a directory onto the stack
    pub fn push(&self, dir: &str) -> Result<(), StackError> {
        let mut entries = self.load()?;
        entries.push(dir.to_string());
        self.save(&entries)
    }

    /// Pop and return the top directory from the stack
    pub fn pop(&self) -> Result<String, StackError> {
        let mut entries = self.load()?;

        if entries.is_empty() {
            return Err(StackError::Empty);
        }

        let dir = entries.pop().unwrap();
        self.save(&entries)?;
        Ok(dir)
    }

    /// Peek at the top directory without removing it
    pub fn peek(&self) -> Result<String, StackError> {
        let entries = self.load()?;

        entries.last().cloned().ok_or(StackError::Empty)
    }

    /// Get the number of entries in the stack
    pub fn size(&self) -> Result<usize, StackError> {
        Ok(self.load()?.len())
    }

    /// Clear all entries from the stack
    pub fn clear(&self) -> Result<(), StackError> {
        self.save(&[])
    }

    fn load(&self) -> Result<Vec<String>, StackError> {
        if !self.path.exists() {
            return Ok(Vec::new());
        }

        let file = File::open(&self.path)?;
        let reader = BufReader::new(file);
        let mut entries = Vec::new();

        for line in reader.lines() {
            let line = line?;
            let trimmed = line.trim();
            if !trimmed.is_empty() {
                entries.push(trimmed.to_string());
            }
        }

        Ok(entries)
    }

    fn save(&self, entries: &[String]) -> Result<(), StackError> {
        // Ensure parent directory exists
        if let Some(parent) = self.path.parent() {
            fs::create_dir_all(parent)?;
        }

        let mut file = File::create(&self.path)?;
        for entry in entries {
            writeln!(file, "{}", entry)?;
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn test_push_pop() {
        let dir = tempdir().unwrap();
        let stack = Stack::new(dir.path().join("stack"));

        stack.push("/home/user/a").unwrap();
        stack.push("/home/user/b").unwrap();

        assert_eq!(stack.size().unwrap(), 2);
        assert_eq!(stack.pop().unwrap(), "/home/user/b");
        assert_eq!(stack.pop().unwrap(), "/home/user/a");
        assert!(matches!(stack.pop(), Err(StackError::Empty)));
    }

    #[test]
    fn test_peek() {
        let dir = tempdir().unwrap();
        let stack = Stack::new(dir.path().join("stack"));

        stack.push("/home/user").unwrap();
        assert_eq!(stack.peek().unwrap(), "/home/user");
        assert_eq!(stack.size().unwrap(), 1); // Still there
    }

    #[test]
    fn test_clear() {
        let dir = tempdir().unwrap();
        let stack = Stack::new(dir.path().join("stack"));

        stack.push("/a").unwrap();
        stack.push("/b").unwrap();
        stack.clear().unwrap();
        assert_eq!(stack.size().unwrap(), 0);
    }

    #[test]
    fn test_empty_stack_errors() {
        let dir = tempdir().unwrap();
        let stack = Stack::new(dir.path().join("stack"));

        assert!(matches!(stack.pop(), Err(StackError::Empty)));
        assert!(matches!(stack.peek(), Err(StackError::Empty)));
    }

    #[test]
    fn test_persistence() {
        let dir = tempdir().unwrap();
        let stack_path = dir.path().join("stack");

        // Push with one instance
        {
            let stack = Stack::new(stack_path.clone());
            stack.push("/first").unwrap();
            stack.push("/second").unwrap();
        }

        // Read with another instance
        {
            let stack = Stack::new(stack_path);
            assert_eq!(stack.size().unwrap(), 2);
            assert_eq!(stack.peek().unwrap(), "/second");
        }
    }
}
